### ISSUE TITLE
Interpretability: add public attention/Q/K accessors on MultiHeadAttentionLayer; cached state has zero-readback and stale-reference bugs

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1359]
## Summary

Wiring HarmonicEngine's interpretability harness (attention visualization + integrated-gradients saliency) against AiDotNet 0.198.x exposed three independent failure modes when extracting attention state from `MultiHeadAttentionLayer<T>` after `model.Predict(input)`:

1. **`_lastAttentionScores` zero-readback for head index > 0** (and the only head when `heads=1`)
2. **`_lastProjectedQueries` / `_lastProjectedKeys` zero / NaN readback** across Predict calls
3. **`_lastQueryInput` stale-reference NaN** for deeper-layer MHA in multi-layer Transformers

All three caches are private fields — there is no public accessor — so the immediate workaround on the consumer side is `private FieldInfo` reflection, which surfaces these data-corruption issues. The minimum fix is a public accessor that snapshots a fresh contiguous copy at call time.

## Repro

`Phase_PAPER_A_PathB_Interpretability_Sanity` in https://github.com/ooples/HarmonicEngine (branch `research/ledger-overhaul-and-l4-reification`):

- `Transformer<float>` with `dModel=128`, `heads=2`, `L=2`, `ctxLen=64`, `dFf=256`
- 2 epochs of per-sample `model.Train(sample, target)` on WikiText-2 first 10KB
- Single `model.Predict(input)` after `SetTrainingMode(false)`
- Reflect into each `MultiHeadAttentionLayer<float>` and read `_lastAttentionScores` (shape `[1, heads, 64, 64]`)

### Observed

```
Layer 0 head 0 query 0: row sum=1.000000, min=1.087e-2, max=5.051e-2   (correct)
Layer 0 head 1 query 0: row sum=0.000000, min=0.000e0,  max=0.000e0    (all zeros)
```

With `heads=1`, even head 0 reads zeros — i.e. the bug isn't strictly "head > 0", it's a storage-aliasing issue that depends on shape + ordering.

Reading `_lastProjectedQueries` / `_lastProjectedKeys` and recomputing `softmax(QK^T / √d_k)` ourselves works for layer 0 but produces NaN for layer 1 (the cached `_lastQueryInput` tensor for the second MHA appears to alias memory that downstream LayerNorm / FFN layers recycle).

Falling back to reading `_lastQueryInput` and doing the Q/K projection ourselves with `GetQueryWeights()` / `GetKeyWeights()` works for layer 0 (the first MHA's cached input is intact) but layer 1's cached input is corrupted (NaN/Inf).

### Expected

Either:
- (a) The cached `_lastAttentionScores` / `_lastProjectedQueries` / `_lastProjectedKeys` / `_lastQueryInput` should remain valid for the lifetime of the layer instance after a Predict, until the next forward overwrites them. OR
- (b) A public accessor (e.g. `GetLastAttentionWeights()`) snapshots the cache into a fresh contiguous tensor and returns that, decoupling consumer access from upstream pool reuse.

## Proposed fix

Add public accessors mirroring the existing `GetQueryWeights()` / `GetKeyWeights()` pattern:

```csharp
public class MultiHeadAttentionLayer<T> : LayerBase<T>
{
    /// <summary>
    /// Returns a fresh contiguous copy of the last forward pass's post-softmax
    /// attention weight tensor, shape [batch, heads, seqQ, seqKV]. Returns null
    /// if no forward has been run since construction or ResetState. The returned
    /// tensor is independent of the layer's internal cache and survives any
    /// subsequent pool reuse / Predict calls.
    /// </summary>
    public Tensor<T>? GetLastAttentionWeights()
    {
        if (_lastAttentionScores is null) return null;
        // Snapshot into a fresh array so the consumer's tensor is decoupled from
        // any pool-rented storage the upstream SDPA fast path returned.
        return _lastAttentionScores.Contiguous().Clone();
    }

    public Tensor<T>? GetLastProjectedQueries() { ... }
    public Tensor<T>? GetLastProjectedKeys() { ... }
}
```

This unblocks the standard interpretability-tool pattern:

```csharp
foreach (var layer in transformer.Layers)
{
    if (layer is MultiHeadAttentionLayer<float> mha)
    {
        var attn = mha.GetLastAttentionWeights();
        // ... visualize, compute rollout, etc.
    }
}
```

The existing `AttentionVisualizationExplainer<T>` constructor already accepts a `Func<Tensor<T>, int, Tensor<T>>` for this purpose — what's missing is the cached-attention public surface to wire it to.

## Test plan

- [ ] Add public `GetLastAttentionWeights()` / `GetLastProjectedQueries()` / `GetLastProjectedKeys()` to `MultiHeadAttentionLayer<T>` that return `.Clone()`'d contiguous copies
- [ ] Unit test: 2-layer Transformer, `heads=2`, ctx=64; after Predict, all 4 layer×head row-sums equal 1.0 ± 1e-3
- [ ] Unit test: assert `GetLastAttentionWeights()` survives 100 subsequent Predict calls without reading zeros / NaN

Filed by HarmonicEngine interpretability scaffolding work — see https://github.com/ooples/HarmonicEngine/tree/research/ledger-overhaul-and-l4-reification for the test harness.
